syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.examples.helloworld";
option java_outer_classname = "HelloWorldProto";
option objc_class_prefix = "HLW";

package helloworld;

// The greeting service definition.
service TheOther98 {
    rpc GetFeed (FeedRequest) returns (FeedResponseView) {}
    rpc GetPost (Id) returns (PostView) {}
    rpc GetProfile (Handle) returns (Profile) {}
    rpc CreatePost (CreatePostRequest) returns (Result) {}
    rpc CreateComment (Comment) returns (Result) {}
}

message Id {
    string value = 1;
}

message FeedRequest {
    repeated string postTags = 1;
    string pageId = 2;
    int32 pageSize = 3;
}

message Handle {
    string value = 1;
}

message Result {
    StatusCode statusCode = 1;
    string errorMessage = 2;

    enum StatusCode {
        OK = 0;
        UNAUTHORIZED = 1;
        FORBIDDEN = 2;
        NOT_FOUND = 3;
        INTERNAL_ERROR = 4;
    }
}

message PostSmallView {
    string title = 1;
    string description = 2;
    string featuredImageLink = 3;
    //string featuredVideoLink = 4;
    //string featuredCaption = 5;
    int64 createDate = 6;
    string type = 7;
    string authorHandle = 8;
}

message Post {
    PostSmallView postSmallView = 1;
    repeated ContentBlock contentBlocks = 2;
    repeated string postTags = 3;
}

message PostView {
    string id = 1;
    PostSmallView postSmallView = 2;
    repeated ContentBlock contentBlocks = 3;
    repeated string postTags = 4;
    repeated Comment comments = 5;
}

message PostFeedView {
    string postViewId = 1;
    PostSmallView postSmallView = 2;
    int64 numberOfComments = 3;
    int64 dateOfLastComment = 4;
}

message FeedResponseView {
    repeated PostFeedView postFeedViews = 1;
    string nextPageId = 2;
}

message CreatePostRequest {
    Post post = 1;
    repeated string viewable_roles = 2;
}

message Profile {
  string handle = 1;
  //string firstName = 2;
  //string lastName = 3;
  //string description = 4;
  //string phone = 5;
  string email = 6;
  //string gender = 7;
  string type = 8;
  //string theme = 9;
}

message ContentBlock {
    ContentBlockType type = 1;
    string content = 2;
    repeated ContentBlock children = 3;

    enum ContentBlockType {
      Text = 0;
      Image = 1;
      ImageGallery = 2;
      Link = 3;
      Video = 4;
      Quote = 5;
    }
}

message Comment {
    int64 id = 1;
    string postViewId = 2;
    int64 createDateMillis = 3;
    string authorHandle = 4;
    //repeated ContentBlock contentBlocks = 5;
    string text = 5;
    int64 score = 6;
}

message CommentView {
    Comment comment = 1;
    int32 userVote = 3;
}